/*
  File: img_base.proto
  Date: May, 2018
  Author: Dagim Sisay <dagiopia@gmail.com>
  License: AGPL
*/

syntax = "proto3";

package ImageBase;

service ImageServices {
  rpc SalientPoint (Image) returns (Point) {}
  rpc ImageStreamPoint (stream Image) returns (Point) {}

  rpc FaceLandmarks (Image) returns (Landmarks) {}
  rpc ImageStreamPoints (stream Image) returns (Points) {}
  
  rpc ImageRect (Image) returns (Rect) {}
  rpc ImageStreamRect (stream Image) returns (Rect) {} 

  rpc DetectFaces (Image) returns (Faces) {}
  rpc ImageStreamRects (stream Image) returns (Rects) {}

  rpc ImageCirc (Image) returns (Circ) {}
}


message Image {
  bytes data = 1;
  int32 operation = 2;
}

message Point {
  int32 x = 1;
  int32 y = 2;
}

message Rect {
  int32 x = 1;
  int32 y = 2;
  int32 w = 3;
  int32 h = 4;
}

message Circ {
  int32 x = 1;
  int32 y = 2;
  float r = 3;
}

message Rects {
  repeated Rect rects = 1;
}

message Points {
  repeated Point points = 1;
}

// face boxes
message Faces {
  repeated Rect faces = 1;
}

// landmarks of multiple faces
message Landmarks {
  repeated Points landmarks = 1;
}

// face features
message FaceF {
  bool known = 1;
  string name = 2;
  Rect location = 3;
  Points landmarks = 4;
  bool smiling = 5;
}

// faces features
message FacesF {
  repeated FaceF facesf = 1;
}
